src/modules/navigator/navigator_main.cpp

```diff
@@ -482,36 +484,112 @@ Navigator::run()
 		    (_geofence.getGeofenceAction() != geofence_result_s::GF_ACTION_NONE) &&
 		    (hrt_elapsed_time(&last_geofence_check) > GEOFENCE_CHECK_INTERVAL)) {
 
-			bool inside = _geofence.check(_global_pos, _gps_pos, _home_pos,
-						      home_position_valid());
+			const position_controller_status_s &pos_ctrl_status = _position_controller_status_sub.get();
+
+			matrix::Vector2<double> fence_violation_test_point;
+			geofence_violation_type_u gf_violation_type{};
+			float test_point_bearing;
+			float test_point_distance;
+			float vertical_test_point_distance;
+
+			if (_vstatus.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING) {
```

Comment:
Also require `_local_pos.v_xy_valid` here?

Reply:
This already checks for the GPS being valid via `have_geofence_position_data` , do you think it needs more?

Reply:
Depending on `GF_SOURCE` (GPS vs global position).

Reply:
@dagar 

Looks like GF_SOURCE logic is broken here. 

https://github.com/PX4/PX4-Autopilot/issues/20677